pgxn/neon/neon_lwlc.c

```diff
@@ -0,0 +1,657 @@
+
+	startup_lsn = controlFile->checkPointCopy.redo;
+
+	Assert(startup_lsn != InvalidXLogRecPtr);
+	Assert(LwLsnCache->n_cached_entries == 0);
+
+	LwLsnCache->highWaterMark = startup_lsn;
+	LwLsnCache->lowWaterMark = startup_lsn;
+	LwLsnCache->initialized = true;
+
+	LWLockRelease(LwLsnMetadataLock);
+}
+
+//int lwlc_pheap_comparefunc(const pairingheap_node *a, const pairingheap_node *b, void *arg) {
+//	const LwLsnCacheEntryData *a_entry;
+//	const LwLsnCacheEntryData *b_entry;
+//
+//	a_entry = pairingheap_const_container(LwLsnCacheEntryData, pheapnode, a);
+//	b_entry = pairingheap_const_container(LwLsnCacheEntryData, pheapnode, b);
+//
+//	if (a_entry->lsn < b_entry->lsn)
+//		return -1;
+//	else if (a_entry->lsn > b_entry->lsn)
+//		return 1;
+//	else
+//		return 0;
+//}
+
+XLogRecPtr
+lwlc_get_max_evicted(void)
+{
+	return LwLsnCache->highWaterMark;
+}
+
+
+static XLogRecPtr
+lwlc_lookup_last_lsn(LwLsnCacheRequest lookup_type, RelFileNode node,
+					 ForkNumber fork, BlockNumber blkno,
+					 XLogRecPtr *effective)
+{
+	uint32		block_hash,
+				relfork_hash,
+				db_hash,
+				cluster_hash;
+	bool		found;
+	XLogRecPtr	lsn;
+	XLogRecPtr	local_effective;
+	LwLsnCacheEntryKeyData key = {0};
+	LwLsnCacheEntryData *entry;
+
+	if (lsn_cache_size <= 0)
+		return InvalidXLogRecPtr;
+
+	/*
+	 * By now we should have run recovery initialization in the startup process,
+	 * so this must be initialized.
+	 */
+	Assert(LwLsnCache->initialized);
+
+	switch (lookup_type) {
+	case LWLSN_BLOCK:
+		MemSet(&key, 0, sizeof(LwLsnCacheEntryKeyData));
+		key.blockNum = blkno;
+		key.rnode = node;
+		key.forkNum = fork;
+		block_hash = get_hash_value(LwLsnCacheTable, &key);
+		/* fallthrough */
+	case LWLSN_RELFORK:
+		MemSet(&key, 0, sizeof(LwLsnCacheEntryKeyData));
+		key.rnode = node;
+		key.forkNum = fork;
+		key.blockNum = InvalidBlockNumber;
+		relfork_hash = get_hash_value(LwLsnCacheTable, &key);
+		/* fallthrough */
+	case LWLSN_DB:
+		MemSet(&key, 0, sizeof(LwLsnCacheEntryKeyData));
+		key.rnode.dbNode = node.dbNode;
+		key.rnode.spcNode = InvalidOid;
+		key.rnode.relNode = InvalidOid;
+		key.blockNum = InvalidBlockNumber;
+		key.forkNum = InvalidForkNumber;
+		db_hash = get_hash_value(LwLsnCacheTable, &key);
+		/* fallthrough */
+	case LWLSN_CLUSTER:
+		MemSet(&key, 0, sizeof(LwLsnCacheEntryKeyData));
+		key.rnode.dbNode = InvalidOid;
+		key.rnode.spcNode = InvalidOid;
+		key.rnode.relNode = InvalidOid;
+		key.blockNum = InvalidBlockNumber;
+		key.forkNum = InvalidForkNumber;
+		cluster_hash = get_hash_value(LwLsnCacheTable, &key);
+		/* fallthrough */
+	}
+
+	LWLockAcquire(LwLsnMetadataLock, LW_SHARED);
+
+	entry = (LwLsnCacheEntryData *)
+		hash_search_with_hash_value(LwLsnCacheTable, &key, block_hash,
+									HASH_FIND,
+									NULL);
+
+	switch (lookup_type) {
+	case LWLSN_BLOCK:
+		MemSet(&key, 0, sizeof(LwLsnCacheEntryKeyData));
+		key.blockNum = blkno;
+		key.rnode = node;
+		key.forkNum = fork;
+		entry = (LwLsnCacheEntryData *)
```

Comment:
Should we perform this lookup after switch to avoid code duplication: all switch branches contain the same call of hash_search_with_hash_value